pkg/cloudprovider/providers/vsphere/vsphere_util.go

```diff
@@ -190,7 +195,90 @@ func (vs *VSphere) getDatastoreMo(ctx context.Context, hubs []pbmtypes.PbmPlacem
 }
 
 // Get all datastores accessible for the virtual machine object.
-func (vs *VSphere) getAllAccessibleDatastores(ctx context.Context, vmObj *object.VirtualMachine) ([]types.ManagedObjectReference, error) {
+func (vs *VSphere) getSharedDatastoresInK8SCluster(ctx context.Context) ([]types.ManagedObjectReference, error) {
+	f := find.NewFinder(vs.client.Client, true)
+	dc, err := f.Datacenter(ctx, vs.cfg.Global.Datacenter)
+	f.SetDatacenter(dc)
+	vmFolder, err := f.Folder(ctx, strings.TrimSuffix(vs.cfg.Global.WorkingDir, "/"))
+	if err != nil {
+		return nil, err
+	}
+	vmMoList, err := vs.GetVMsInsideFolder(ctx, vmFolder, []string{NameProperty})
+	if err != nil {
+		return nil, err
+	}
+	dsMorefs := make(map[int][]types.ManagedObjectReference)
```

Comment:
Comment on line#210-219

This entire block can be written in a simple and efficient way. There is no need to create multiple maps to arrive at the shared datastore list and also I found it a bit complex to understand the logic. Here's how we can write this block (pseudo code only):

Set<String> sharedDs;
for (i = 0; i < vmList.size(); i++) {
	Set<String> accessibleDs = getAccesibleDatastoresForNode(vm.get(i))
	if (i == 0) {
		sharedDs.addAll(accesibleDs);
	} else {
		sharedDs = intersect(sharedDs, accessibleDs);
		if (sharedDs.size() == 0) {
			break;
		}
	}
}

sharedDs intersect(sharedDs, accessibleDs) {
	for ds : sharedDs {
		if (accessibleDs.get(ds) == null) {
			sharedDs.remove(ds)
		}
	}
	return sharedDs
}